kv.go

```diff
@@ -0,0 +1,436 @@
+		Description:       cfg.Description,
+		Subjects:          []string{fmt.Sprintf(kvSubjectsTmpl, cfg.Bucket)},
+		MaxMsgsPerSubject: history,
+		MaxBytes:          cfg.MaxBytes,
+		MaxAge:            cfg.TTL,
+		MaxMsgSize:        cfg.MaxValueSize,
+		Storage:           cfg.Storage,
+		Replicas:          replicas,
+	}
+
+	if _, err := js.AddStream(scfg); err != nil {
+		return nil, err
+	}
+
+	kv := &kvs{
+		name:   cfg.Bucket,
+		stream: scfg.Name,
+		pre:    fmt.Sprintf(kvSubjectsPreTmpl, cfg.Bucket),
+		js:     js,
+	}
+	return kv, nil
+}
+
+// DeleteKeyValue will delete this KeyValue store (JetStream stream).
+func (js *js) DeleteKeyValue(bucket string) error {
+	stream := fmt.Sprintf(kvBucketNameTmpl, bucket)
+	return js.DeleteStream(stream)
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	js     *js
+}
+
+// Underlying entry.
+type kve struct {
+	bucket   string
+	key      string
+	value    []byte
+	revision uint64
+	delta    uint64
+	created  time.Time
+	op       KeyValueOp
+}
+
+func (e *kve) Bucket() string        { return e.bucket }
+func (e *kve) Key() string           { return e.key }
+func (e *kve) Value() []byte         { return e.value }
+func (e *kve) Revision() uint64      { return e.revision }
+func (e *kve) Created() time.Time    { return e.created }
+func (e *kve) Delta() uint64         { return e.delta }
+func (e *kve) Operation() KeyValueOp { return e.op }
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (KeyValueEntry, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m, err := kv.js.GetLastMsg(kv.stream, b.String())
+	if err != nil {
+		return nil, err
+	}
+
+	entry := &kve{
+		bucket:   kv.name,
+		key:      key,
+		value:    m.Data,
+		revision: m.Sequence,
+		created:  m.Time,
+	}
+
+	// Double check here that this is not a DEL Operation marker.
+	if len(m.Header) > 0 && m.Header.Get(kvop) == kvdel {
+		entry.op = KeyValueDelete
+		return entry, ErrMSgNotFound
+	}
+
+	return entry, nil
+}
+
+// Put will place the new value for the key into the store.
+func (kv *kvs) Put(key string, value []byte) (revision uint64, err error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	pa, err := kv.js.Publish(b.String(), value)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Create will add the key/value pair iff it does not exist.
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	v, err := kv.Update(key, value, 0)
+	if err == nil {
+		return v, nil
+	}
+	// TODO(dlc) - Since we have tombstones for DEL ops for watchers, this could be from that
+	// so we need to double check.
+	if e, err := kv.Get(key); err == ErrMSgNotFound {
+		return kv.Update(key, value, e.Revision())
+	}
+	return 0, err
+}
+
+// Update will update the value iff the latest revision matches.
+func (kv *kvs) Update(key string, value []byte, revision uint64) (uint64, error) {
```

Comment:
The godoc should clearly indicate these deviations from the spec if you wish to keep them - and keep adding more functions over time - vs the options to put approach.  And the godoc should provide a indication that in other implementation if you did that now you do this etc

Reply:
Which spec are you referring to, the ADR?

Help me understand what you are looking for here.

Reply:
yeah, ADR would have `Put(k, v, opts...)` to implement Update and Creates and whatever other things we wish to add in future. Though I agree Creates() reads nicer, so feel free to keep them instead, just point out in the comments that this is equivalent to `Put(k, v, OnlyIfLastValueSequence(0))` though that's not nice think we're still thinking about actual correct names.

Reply:
Gotcha.